k2/python/csrc/torch/fsa_algo.cu

```diff
@@ -0,0 +1,101 @@
+/**
+ * @brief python wrappers for fsa_algo.
+ *
+ * @copyright
+ * Copyright (c)  2020  Mobvoi Inc.        (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include <utility>
+#include <vector>
+
+#include "k2/csrc/fsa.h"
+#include "k2/csrc/fsa_algo.h"
+#include "k2/python/csrc/torch/fsa_algo.h"
+#include "k2/python/csrc/torch/torch_util.h"
+
+namespace k2 {
+
+static void PybindTopSort(py::module &m) {
+  // TODO(fangjun): add docstring for this function
+  //
+  // if need_arc_map is true, it returns (sorted_fsa_vec, arc_map);
+  // otherwise, it returns (sorted_fsa_vec, None)
+  m.def(
+      "top_sort",
+      [](FsaVec &src, bool need_arc_map = true)
+          -> std::pair<FsaVec, torch::optional<torch::Tensor>> {
+        Array1<int32_t> arc_map;
+        FsaVec sorted;
+        TopSort(src, &sorted, need_arc_map ? &arc_map : nullptr);
+        torch::optional<torch::Tensor> tensor;
+        if (need_arc_map) tensor = ToTensor(arc_map);
+        return std::make_pair(sorted, tensor);
+      },
+      py::arg("src"), py::arg("need_arc_map") = true);
+}
+
+static void PybindLinearFsa(py::module &m) {
+  // TODO(fangjun): make std::vector an opaque type for Pybind11
+  // to avoid copying data from Python to C++.
+  m.def(
+      "linear_fsa",
+      [](const std::vector<int32_t> &symbols, int32_t gpu_id = -1) -> Fsa {
```

Comment:
So what kind of Python objects will get converted to std::vector?  
I was thinking this might be a Tensor, which could be converted to Array1 with zero-copy.

Reply:
A python list object [1, 2, 3] is **copied** to construct a `std::vector<int>` in C++.

A Python list object [ [1, 2, 3], [4, 5, 6] ] is **copied** to construct a `std::vector<std::vector<int>>` in C++.

---

Using a `torch::Tensor` as arguments involves NO memory copy.

For a list of list of integers, can we pass two 1-D tensors:
- one with the integer ids
- another with size information about the first tensor?